staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1201,41 +1294,77 @@ func (c *cacheWatcher) ResultChan() <-chan watch.Event {
 func (c *cacheWatcher) stopLocked() {
 	if !c.stopped {
 		c.stopped = true
-		close(c.done)
 		close(c.input)
 	}
+
+	// stop without draining the input channel was requested
+	// make sure the done channel is not already closed otherwise it will panic
+	if !c.drainInputBuffer && !c.isDoneChannelClosed() {
+		close(c.done)
+	}
 }
 
 func (c *cacheWatcher) nonblockingAdd(event *watchCacheEvent) bool {
+	// if the bookmarkAfterResourceVersion hasn't been seen
+	// we will try to deliver a bookmark event every second.
+	// the following check will discard a bookmark event
+	// if it is < than the bookmarkAfterResourceVersion
+	// so that we don't pollute the input channel
+	if event.Type == watch.Bookmark && event.ResourceVersion < c.bookmarkAfterResourceVersion {
+		return false
+	}
 	select {
 	case c.input <- event:
+		c.markBookmarkAfterResourceVersionAsReceived(event)
 		return true
 	default:
 		return false
 	}
 }
 
 // Nil timer means that add will not block (if it can't send event immediately, it will break the watcher)
+//
+// Note that bookmark events are never added via the add method only via the nonblockingAdd.
+// Changing this behaviour will require moving the markBookmarkAfterResourceVersionAsReceived method
 func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer) bool {
 	// Try to send the event immediately, without blocking.
 	if c.nonblockingAdd(event) {
 		return true
 	}
 
 	closeFunc := func() {
-		// This means that we couldn't send event to that watcher.
-		// Since we don't want to block on it infinitely,
-		// we simply terminate it.
-		klog.V(1).Infof("Forcing %v watcher close due to unresponsiveness: %v. len(c.input) = %v, len(c.result) = %v", c.objectType.String(), c.identifier, len(c.input), len(c.result))
 		metrics.TerminatedWatchersCounter.WithLabelValues(c.objectType.String()).Inc()
-		c.forget()
+		// This means that we couldn't send event to that watcher.
+		// Since we don't want to block on it infinitely, we simply terminate it.
+
+		// we are graceful = false, when:
+		//
+		// (a) The bookmarkAfterResourceVersionReceived hasn't been received,
+		//     we can safely terminate the watcher. Because the client is waiting
+		//     for this specific bookmark, and we even haven't received one.
+		// (b) We have seen the bookmarkAfterResourceVersion, and it was sent already to the client.
+		//     We can simply terminate the watcher.
+
+		// we are graceful = true, when:
+		//
+		// (a) We have seen a bookmark, but it hasn't been sent to the client yet.
+		//     That means we should drain the input buffer which contains
+		//     the bookmarkAfterResourceVersion we want. We do that to make progress
+		//     as clients can re-establish a new watch with the given RV and receive
+		//     further notifications.
+
+		c.stateMutex.Lock()
+		defer c.stateMutex.Unlock()
+		graceful := c.wasBookmarkAfterResourceVersionReceivedLocked() && !c.wasBookmarkAfterResourceVersionSent()
```

Comment:
Let's switch to:
```
graceful := (c.state == cacheWatcherBookmarkReceived)
```

Seems easier to follow and understand (especially given you take the lock explicitly here anyway.

Reply:
Also switch to:
```
graceful := func() {
  c.stateMute.Lock()
  defer c.stateMutex.Unlock()
  return c.state == cacheWatcherBookmarkReceived
}()
```

[avoids locking for the time of calling forget]

Reply:
we are graceful when we have received an RV but i hasn't been sent, so we cannot simply check `graceful := (c.state == cacheWatcherBookmarkReceived)`